File: cmdb-web/backend/app/api/routes/audit.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.audit import AuditLog
from app.schemas.common import BaseResponse

router = APIRouter()


@router.get("", response_model=BaseResponse)
async def list_audit_logs(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    username: str | None = None,
    action: str | None = None,
    resource_type: str | None = None,
    status: str | None = None,
    start_date: str | None = None,
    end_date: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    query = select(AuditLog)
    count_query = select(func.count(AuditLog.id))

    if username:
        query = query.where(AuditLog.username == username)
        count_query = count_query.where(AuditLog.username == username)
    if action:
        query = query.where(AuditLog.action == action)
        count_query = count_query.where(AuditLog.action == action)
    if resource_type:
        query = query.where(AuditLog.resource_type == resource_type)
        count_query = count_query.where(AuditLog.resource_type == resource_type)
    if status:
        query = query.where(AuditLog.status == status)
        count_query = count_query.where(AuditLog.status == status)
    if start_date:
        try:
            start_dt = datetime.fromisoformat(start_date)
            query = query.where(AuditLog.created_at >= start_dt)
            count_query = count_query.where(AuditLog.created_at >= start_dt)
        except ValueError:
            pass
    if end_date:
        try:
            end_dt = datetime.fromisoformat(end_date)
            query = query.where(AuditLog.created_at <= end_dt)
            count_query = count_query.where(AuditLog.created_at <= end_dt)
        except ValueError:
            pass

    total_result = await db.execute(count_query)
    total = total_result.scalar() or 0

    query = query.offset((page - 1) * page_size).limit(page_size).order_by(AuditLog.created_at.desc())
    result = await db.execute(query)
    items = result.scalars().all()

    action_map = {
        "create": "创建", "update": "更新", "delete": "删除",
        "login": "登录", "logout": "登出", "export": "导出",
        "import": "导入", "approve": "审批", "reject": "拒绝",
    }
    resource_type_map = {
        "ci": "配置项", "change": "变更", "user": "用户",
        "role": "角色", "permission": "权限", "system": "系统",
    }

    log_list = []
    for log in items:
        log_list.append({
            "id": log.id,
            "user": log.username,
            "userId": log.user_id,
            "action": log.action,
            "actionName": action_map.get(log.action, log.action),
            "resource": log.resource_name,
            "resourceType": resource_type_map.get(log.resource_type, log.resource_type or ""),
            "resourceId": log.resource_id,
            "ip": log.ip,
            "userAgent": log.user_agent,
            "status": log.status,
            "details": log.details,
            "oldValue": log.old_value,
            "newValue": log.new_value,
            "createdAt": log.created_at.strftime("%Y-%m-%d %H:%M:%S") if log.created_at else "",
        })

    return BaseResponse(data={"items": log_list, "total": total})


@router.get("/export", response_model=BaseResponse)
async def export_audit_logs(
    username: str | None = None,
    action: str | None = None,
    resource_type: str | None = None,
    start_date: str | None = None,
    end_date: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    query = select(AuditLog)

    if username:
        query = query.where(AuditLog.username == username)
    if action:
        query = query.where(AuditLog.action == action)
    if resource_type:
        query = query.where(AuditLog.resource_type == resource_type)
    if start_date:
        try:
            start_dt = datetime.fromisoformat(start_date)
            query = query.where(AuditLog.created_at >= start_dt)
        except ValueError:
            pass
    if end_date:
        try:
            end_dt = datetime.fromisoformat(end_date)
            query = query.where(AuditLog.created_at <= end_dt)
        except ValueError:
            pass

    query = query.order_by(AuditLog.created_at.desc()).limit(1000)
    result = await db.execute(query)
    items = result.scalars().all()

    export_data = []
    for log in items:
        export_data.append({
            "id": log.id,
            "username": log.username,
            "action": log.action,
            "resourceType": log.resource_type,
            "resourceName": log.resource_name,
            "ip": log.ip,
            "status": log.status,
            "createdAt": log.created_at.strftime("%Y-%m-%d %H:%M:%S") if log.created_at else "",
        })

    return BaseResponse(data={"items": export_data, "total": len(export_data)})

File: cmdb-web/backend-go/internal/routes/audit.go
package routes

import (
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"cmdb-go/internal/database"
	"cmdb-go/internal/models"
	"cmdb-go/internal/schemas"
)

// ListAuditLogs 获取审计日志列表
func ListAuditLogs(c *gin.Context) {
	page := 1
	pageSize := 20
	if p := c.Query("page"); p != "" {
		page, _ = strconv.Atoi(p)
	}
	if ps := c.Query("pageSize"); ps != "" {
		pageSize, _ = strconv.Atoi(ps)
	}
	if ps := c.Query("page_size"); ps != "" {
		pageSize, _ = strconv.Atoi(ps)
	}
	if page <= 0 {
		page = 1
	}
	if pageSize <= 0 {
		pageSize = 20
	}

	keyword := c.Query("keyword")
	action := c.Query("action")
	status := c.Query("status")
	startTime := c.Query("start_time")
	endTime := c.Query("end_time")

	db := database.GetDB()
	var total int64

	query := db.Model(&models.AuditLog{})
	if keyword != "" {
		query = query.Where("username LIKE ? OR resource_name LIKE ? OR resource_id LIKE ?", "%"+keyword+"%", "%"+keyword+"%", "%"+keyword+"%")
	}
	if action != "" {
		query = query.Where("action = ?", action)
	}
	if status == "failed" {
		// 当前表结构没有独立状态字段，约定通过 action=failed 兼容筛选能力
		query = query.Where("action = ?", "failed")
	}
	if status == "success" {
		query = query.Where("action <> ?", "failed")
	}
	if startTime != "" {
		if t, err := time.Parse("2006-01-02 15:04:05", startTime); err == nil {
			query = query.Where("created_at >= ?", t)
		}
	}
	if endTime != "" {
		if t, err := time.Parse("2006-01-02 15:04:05", endTime); err == nil {
			query = query.Where("created_at <= ?", t)
		}
	}
	query.Count(&total)

	var logs []models.AuditLog
	query.Offset((page-1)*pageSize).Limit(pageSize).Order("created_at DESC").Find(&logs)

	c.JSON(http.StatusOK, schemas.BaseResponse{
		Success: true,
		Message: "ok",
		Data: gin.H{
			"items": logs,
			"total": total,
			"page": page,
			"page_size": pageSize,
		},
	})
}
